File: frontend/lib/api/verification.ts
import type { ProjectPackage } from "@/lib/api/packaging";
import type { ProjectWorkspace } from "@/lib/api/workspace";

export type VerificationStatus = "passed" | "failed";

export type SandboxRun = {
  build_status: VerificationStatus;
  exit_code: number;
  finished_at: string;
  package_id: string;
  started_at: string;
  status: VerificationStatus;
  test_status: VerificationStatus;
  verification_id: string;
  verification_summary: string;
};

export type BuildResult = {
  build_logs: string[];
  exit_code: number;
  failed_checks: number;
  passed_checks: number;
  status: VerificationStatus;
  target: string;
};

export type VerificationReport = {
  build_results: BuildResult[];
  sandbox_run: SandboxRun;
  source_package_id: string;
  source_workspace_id: string;
  source_workspace_updated_at: string;
};

export type ProjectVerificationInput = {
  packageIncludedFiles: string[];
  projectPackage: ProjectPackage;
  workspace: ProjectWorkspace;
};

type ApiErrorPayload = { code?: string; detail?: string; message?: string };

export class ProjectVerificationApiError extends Error {
  readonly code: string;
  readonly status: number;

  constructor({
    code,
    message,
    status,
  }: Readonly<{ code: string; message: string; status: number }>) {
    super(message);
    this.name = "ProjectVerificationApiError";
    this.code = code;
    this.status = status;
  }
}

const apiBaseUrl = process.env.NEXT_PUBLIC_API_BASE_URL ?? "http://localhost:8000";

/** Performs deterministic, non-executing structural verification of an existing package. */
export async function requestProjectVerification(
  input: Readonly<ProjectVerificationInput>,
): Promise<VerificationReport> {
  let response: Response;

  try {
    response = await fetch(`${apiBaseUrl}/api/v1/verifications`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        package_included_files: input.packageIncludedFiles,
        project_package: input.projectPackage,
        workspace: input.workspace,
      }),
    });
  } catch {
    throw new ProjectVerificationApiError({
      code: "verification_unreachable",
      message: "Genesis could not reach the Project Verification Engine. Please try again.",
      status: 0,
    });
  }

  if (!response.ok) {
    const payload = (await response.json().catch(() => ({}))) as ApiErrorPayload;
    throw new ProjectVerificationApiError({
      code: payload.code ?? "verification_error",
      message: payload.message ?? payload.detail ?? "Genesis could not verify the project.",
      status: response.status,
    });
  }

  return (await response.json()) as VerificationReport;
}

File: backend/app/schemas/verification.py
from datetime import datetime
from typing import Annotated, Literal

from pydantic import Field

from app.schemas.architect import LongText, ShortText
from app.schemas.base import Schema
from app.schemas.packaging import ProjectPackage
from app.schemas.workspace import ProjectWorkspace

VerificationStatus = Literal["passed", "failed", "pending"]
ProjectImplementationLevel = Literal["foundation", "partial", "complete"]
VerificationLog = Annotated[str, Field(min_length=1, max_length=1200)]


class SandboxRun(Schema):
    """One safe, deterministic verification run for a packaged project revision."""

    verification_id: ShortText
    package_id: ShortText
    started_at: datetime
    finished_at: datetime
    status: VerificationStatus
    build_status: VerificationStatus
    test_status: VerificationStatus
    implementation_level: ProjectImplementationLevel
    exit_code: Annotated[int, Field(ge=0)]
    verification_summary: LongText


class BuildResult(Schema):
    """One deterministic verification result for a project surface."""

    target: ShortText
    status: VerificationStatus
    exit_code: Annotated[int, Field(ge=0)]
    passed_checks: Annotated[int, Field(ge=0)]
    pending_checks: Annotated[int, Field(ge=0)]
    failed_checks: Annotated[int, Field(ge=0)]
    build_logs: Annotated[list[VerificationLog], Field(min_length=1, max_length=16)]


class VerificationReport(Schema):
    """The aggregate sandbox result, without executing generated user code."""

    sandbox_run: SandboxRun
    build_results: Annotated[list[BuildResult], Field(min_length=1, max_length=8)]
    source_workspace_id: ShortText
    source_workspace_updated_at: datetime
    source_package_id: ShortText


class ProjectVerificationRequest(Schema):
    """Existing workspace and package metadata required for safe deterministic verification."""

    workspace: ProjectWorkspace
    project_package: ProjectPackage
    package_included_files: Annotated[list[ShortText], Field(max_length=300)]
